src/lambdaisland/witchcraft/citizens/trait.clj

```diff
@@ -0,0 +1,228 @@
+  ([this f] (.swap (.state this) f))
+  ([this f arg] (.swap (.state this) f arg))
+  ([this f arg1 arg2] (.swap (.state this) f arg1 arg2))
+  ([this f x y args] (.swap (.state this) f x y args)))
+
+(defn trait-compareAndSet [this oldv newv]
+  (.compareAndSet (.state this) oldv newv))
+
+(defn trait-reset [this newval]
+  (.reset (.state this) newval))
+
+(defn trait-swapVals
+  ([this f] (.swapVals (.state this) f))
+  ([this f arg] (.swapVals (.state this) f arg))
+  ([this f arg1 arg2] (.swapVals (.state this) f arg1 arg2))
+  ([this f x y args] (.swapVals (.state this) f x y args)))
+
+(defn trait-resetVals [this newval]
+  (.resetVals (.state this) newval))
+
+(defn make-trait
+  "Create a subclass of npcs.citizensapi.trait.Trait, `name` is a simple lowercase
+  identifier of the trait like `\"sneak\"` or `\"aggressive\"`. `callbacks` is a
+  map from keyword to function (or var, IFn). Callbacks receive `this` (the
+  trait) as first arg. `:load` and `:save` also receive a `DataKey`.
+
+  Callbacks: `:load`, `:save`, `:on-attach`, `:on-copy`, `:on-despawn`,
+  `:on-pre-spawn`, `:run`, `:save`.
+
+  Can be called multiple times, subsequent calls will only replace the
+  callbacks. Returns the trait subclass `java.lang.Class`.
+
+  Note that adding a `:run` callback for a Trait that is already attached to an
+  NPC will not cause it to be called. Remove the Trait from the NPC and add it
+  back.
+
+  To give your trait state, provide an `:init` key (map or function that returns
+  a map), this will provide your trait's initial state. You can use the trait
+  instance itself as an atom, e.g. call `@this` or `(swap! this ...)` to get/set
+  values. Any state set this way will automatically be persisted by the default
+  `:load`/`:save` implementations, unless the value has a `:no-persist`
+  metadata.
+
+  To initialize the Trait instance after instantiation, use `:post-init`.
+  "
+  [name callbacks]
+  (let [name (str/lower-case name) ; Trait does the same, stay consistent
+        init-fn-name (str "init-" name)
+        classname (str *ns* "." name)]
+    (swap! registry assoc name callbacks)
+    ;; Make a unique `init` function for this class, so we have a way to inject
+    ;; the correct name into the superclass constructor
+    (intern (the-ns (symbol (namespace `_)))
+            (symbol (str "trait-" init-fn-name))
+            (fn [] [[name] (atom (if-let [init (:init callbacks)]
+                                   (if (map? init)
+                                     init
+                                     (init))
+                                   {}))]))
+    (try
+      (Class/forName classname)
+      (catch java.lang.ClassNotFoundException _
+        (gen-and-load-class
+         {:name classname
+          :extends Trait
+          :implements [clojure.lang.IDeref clojure.lang.IAtom2]
+          :state "state"
+          :init init-fn-name
+          :constructors {[] [String]}
+          :post-init `trait-post-init
```

Comment:
```suggestion
          :post-init "post-init"
```

Otherwise I get:
```
Execution error (ClassFormatError) at java.lang.ClassLoader/defineClass1 (ClassLoader.java:-2).
Illegal field name "lambdaisland.witchcraft.citizens.trait_SLASH_trait_post_init__var" in class my/name/space
```
when calling `make-trait`.

Reply:
Yeah sorry I hadn't pushed my code in a while. I pushed my latest changes, including two example implementations of traits.